reactor-core/src/main/java/reactor/core/publisher/Flux.java

```diff
@@ -7231,9 +7237,66 @@ public int getPrefetch() {
+				.map(Retry.RetrySignal::failure)
+				.as(whenFactory)));
+	}
+
+	/**
+	 * Retries this {@link Flux} in response to signals emitted by a companion {@link Publisher}.
+	 * The companion is generated by a {@link Retry} instance supplied through the given {@link Supplier},
+	 * see {@link Retry#max(long)}, {@link Retry#maxInARow(long)} and {@link Retry#backoff(long, Duration)}
+	 * for provided builders.
+	 * <p>
+	 * The operator generates a base for the companion, a {@link Flux} of {@link reactor.util.retry.Retry.RetrySignal}
+	 * which each give metadata about each retryable failure whenever this {@link Flux} signals an error. The final companion
+	 * should be derived from that base companion and emit data in response to incoming onNext (although it can emit less
+	 * elements, or delay the emissions).
+	 * <p>
+	 * Terminal signals in the companion terminate the sequence with the same signal, so emitting an {@link Subscriber#onError(Throwable)}
+	 * will fail the resulting {@link Flux} with that same error.
+	 * <p>
+	 * <img class="marble" src="doc-files/marbles/retryWhenForFlux.svg" alt="">
+	 * <p>
+	 * Note that the {@link Retry.RetrySignal} state can be transient and change between each source
+	 * {@link org.reactivestreams.Subscriber#onError(Throwable) onError} or
+	 * {@link org.reactivestreams.Subscriber#onNext(Object) onNext}. If processed with a delay,
+	 * this could lead to the represented state being out of sync with the state at which the retry
+	 * was evaluated. Map it to {@link Retry.RetrySignal#copy()} right away to mediate this.
+	 * <p>
+	 * Note that if the companion {@link Publisher} created by the {@code whenFactory}
+	 * emits {@link Context} as trigger objects, these {@link Context} will be merged with
+	 * the previous Context:
+	 * <pre><code>
+	 * .retryWhen(errorCurrentAttempt -> errorCurrentAttempt.handle((retrySignal, sink) -> {
+	 * 	    Context ctx = sink.currentContext();
+	 * 	    int rl = ctx.getOrDefault("retriesLeft", 0);
+	 * 	    if (rl > 0) {
+	 *		    sink.next(Context.of(
+	 *		        "retriesLeft", rl - 1,
+	 *		        "lastError", retrySignal.failure()
+	 *		    ));
+	 * 	    } else {
+	 * 	        sink.error(new IllegalStateException("retries exhausted", retrySignal.failure()));
+	 * 	    }
+	 * }))
+	 * </code></pre>
+	 *
+	 * @param retrySpecSupplier the {@link Supplier} of the {@link Retry} strategy that will generate the companion {@link Publisher},
+	 * given a {@link Flux} that signals each onError as a {@link reactor.util.retry.Retry.RetrySignal}.
+	 *
+	 * @return a {@link Flux} that retries on onError when a companion {@link Publisher} produces an onNext signal
+	 * @see Retry#max(long)
+	 * @see Retry#maxInARow(long)
+	 * @see Retry#backoff(long, Duration)
+	 */
+	public final Flux<T> retryWhen(Supplier<Retry> retrySpecSupplier) {
+		return onAssembly(new FluxRetryWhen<>(this, retrySpecSupplier.get()));
```

Comment:
Could you please elaborate why we call `Supplier#get()` here?

Reply:
`Supplier` is just a hack to avoid compiler ambiguity around `x -> y` functional interfaces. thanks to it we've got `() -> x` function shape vs `x -> y` function shape.